streaming/protocol/setup.c

```diff
@@ -0,0 +1,283 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "setup.h"
+#include "handshake.h"
+
+static const size_t HANDSHAKE_PROTOCOL_INITIAL_RESPONSE_SIZE = 1024;
+
+static void rrdpush_set_flags_to_newest_stream(RRDHOST *host) {
+    rrdhost_flag_set(host, RRDHOST_FLAG_STREAM_LABELS_UPDATE);
+    rrdhost_flag_clear(host, RRDHOST_FLAG_STREAM_LABELS_STOP);
+}
+
+static long int parse_stream_version_for_errors(const char *http)
+{
+    if (!memcmp(http, START_STREAMING_ERROR_SAME_LOCALHOST, sizeof(START_STREAMING_ERROR_SAME_LOCALHOST)))
+        return -2;
+    else if (!memcmp(http, START_STREAMING_ERROR_ALREADY_STREAMING, sizeof(START_STREAMING_ERROR_ALREADY_STREAMING)))
+        return -3;
+    else if (!memcmp(http, START_STREAMING_ERROR_NOT_PERMITTED, sizeof(START_STREAMING_ERROR_NOT_PERMITTED)))
+        return -4;
+    else
+        return -1;
+}
+
+static long int parse_stream_version(RRDHOST *host, const char *http)
+{
+    long int stream_version = -1;
+    int answer = -1;
+    char *stream_version_start = strchr(http, '=');
+    if (stream_version_start) {
+        stream_version_start++;
+        stream_version = strtol(stream_version_start, NULL, 10);
+        answer = memcmp(http, START_STREAMING_PROMPT_VN, (size_t)(stream_version_start - http));
+        if (!answer) {
+            rrdpush_set_flags_to_newest_stream(host);
+        }
+    } else {
+        answer = memcmp(http, START_STREAMING_PROMPT_V2, strlen(START_STREAMING_PROMPT_V2));
+        if (!answer) {
+            stream_version = 1;
+            rrdpush_set_flags_to_newest_stream(host);
+        } else {
+            answer = memcmp(http, START_STREAMING_PROMPT, strlen(START_STREAMING_PROMPT));
+            if (!answer) {
+                stream_version = 0;
+                rrdhost_flag_set(host, RRDHOST_FLAG_STREAM_LABELS_STOP);
+                rrdhost_flag_clear(host, RRDHOST_FLAG_STREAM_LABELS_UPDATE);
+            }
+            else {
+                stream_version = parse_stream_version_for_errors(http);
+            }
+        }
+    }
+    return stream_version;
+}
+
+static inline struct netdata_ssl *sender_ssl(struct sender_state *ss) {
+    return &ss->host->ssl;
+}
+
+static inline int sender_sockfd(struct sender_state *ss) {
+    return ss->host->rrdpush_sender_socket;
+}
+
+static inline RRDHOST *receiver_host(struct receiver_state *rs) {
+    return rs->host;
+}
+
+static inline struct netdata_ssl *receiver_ssl(struct receiver_state *rs) {
+    return &rs->ssl;
+}
+
+static inline int receiver_sockfd(struct receiver_state *rs) {
+    return rs->fd;
+}
+
+#define sender_error(sender, fmt, args...) do { \
+    size_t len = 1024; \
+    char newfmt[len]; \
```

Comment:
in normal situation we would use `len +1` for `newfmt`, I will suggest to do the same here to keep pattern.